librosa/core/audio.py

```diff
@@ -141,7 +141,22 @@ def load(path, sr=22050, mono=True, offset=0.0, duration=None,
 
     except RuntimeError as exc:
 
+        # Convert pathlib paths to strings
+        if sys.version_info >= (3, 4):
+            import pathlib
+
+            # check if a Path object is passed by the user
+            if isinstance(path, pathlib.PurePath):
+                # Recommended way of converting to string
+                if sys.version_info >= (3, 6):
+                    path = os.fspath(path)
+                else:
+                    path = str(path)
+
```

Comment:
This looks good, thanks for doing it!

I wonder if we could do something more robust than type checking though, eg always try to cast to str (or fspath) directly even if the input is already a string.  This might circumvent future issues with types that could be coerced into strings, but exist outside the string class hierarchy.  If the cast fails, then we re-raise anyway.

An added bonus of this is that it would work automatically with `Path` objects and not require us to import pathlib here, so it should work across all current (and future) python versions.

Reply:
You're welcome ^-^

If you plan to only support Python 3.4+, you can also reverse that logic. You can always cast it to a Path.
```
path = "folder/foo.wav"
Path(path)  # turns string into Path object
path = Path("folder/foo.wav")
Path(path)  # doesn't do anything, Path object stays Path object
```

Path objects are also stronger in the sense that you can do things like path.exists() (checks if file actually exists), possibly giving stronger checking support.

Conceptually, I would argue against always converting. I think it's better to let the underlying library / function determine whether it can accept it or not, not Librosa itself. PySoundFile already seems to accept Path objects, so converting to strings here only increases the chance of bugs. Also, maybe at some point another library for loading audio might be introduced (e.g. Rust based), which accepts other type of inputs. In that case Librosa's checks need to be updated, which could break / limit future library updates without active involvement of Librosa's developers / community.

I would limit Librosa's checking only to warnings. As in: "Hey, if your audio won't load, that's likely due to a incorrect path / non-existing file / etc, but let's try anyway".

Reply:
Ok, those are fair points.

Regarding future support plans, see #911 for discussion and please chime in if you have thoughts.  Right now, I'm thinking that when we drop 2.7, we'll also drop 3.4 and 3.5 (which will hit EOL not long after our release target of 2020/06 anyway).

Reply:
The only other thing I'd like to see added to this is an expansion of the docstring to include the newly supported type.

Thinking more on your point about always casting: I can see a case for always casting up to Path (rather than down to string), assuming the backend libraries support Path consistently.  They seem to now, and that should only improve in the future, but I'm okay with holding off on that.  My main argument for casting to Path is that it would simplify the code from our side (in terms of explicit type checks and branching points), but I'm not going to be a stickler on it.

Reply:
Where should the docstring be modified? Since `pathlib.Path()` was already supported until `0.7.0`, but only broke for non-wav files in `0.7.1` (it works for properly formatted .wav's, loaded with `sf.SoundFile`), I wouldn't think this is a new feature?

Reply:
It's not a new feature, but the support for Path objects was implicit before, and now it's explicit.  I think the type description on the `path` argument should be expanded to include `..., pathlib.Path, or file-like object`.

Reply:
I think I found the docstring you mentioned. Please check if the updated description is to your liking.

Reply:
That's good, thanks!  I was actually referring to line 47, `path : string, int, or file-like object` but GH won't let me tack a comment directly there.

I've also restarted the broken travis build, hopefully that fixes it.

If you want to amend this one-liner, feel free to stick a `[ci skip]` at the end of your commit line to bypass unit tests, since it's only a doc change at this point.  Otherwise it LGTM.

Reply:
Did the change. I assumed that `file-like object` was already referring to `pathlib.Path`. What's the difference?

Reply:
Sorry, missed this comment before.  A file-like object is one which supports file operations (reading and writing) as documented in the python [io module](https://docs.python.org/3/library/io.html).  This includes things like open file descriptors, but also `BytesIO` etc.

`pathlib.Path` is if anything a string-like object, but you can't read directly from one without opening it first.